internal/sapsystem/sapsystem.go

```diff
@@ -0,0 +1,165 @@
+package sapsystem
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"regexp"
+
+	"github.com/pkg/errors"
+
+	"github.com/SUSE/sap_host_exporter/lib/sapcontrol"
+	"github.com/spf13/viper"
+
+	"github.com/trento-project/trento/internal"
+)
+
+const sapInstallationPath string = "/usr/sap"
+const sapIdentifierPattern string = "^[A-Z][A-Z0-9]{2}$" // PRD, HA1, etc
+const sapInstancePattern string = "^[A-Z]*([0-9]{2})$"   // HDB00, ASCS00, ERS10, etc
+
+type SAPSystemsList []*SAPSystem
+
+type SAPSystem struct {
+	webService sapcontrol.WebService
+	Id         string                                  `mapstructure:"id,omitempty"`
+	Type       string                                  `mapstructure:"type,omitempty"`
+	Processes  map[string]*sapcontrol.OSProcess        `mapstructure:"processes,omitempty"`
+	Instances  map[string]*sapcontrol.SAPInstance      `mapstructure:"instances,omitempty"`
+	Properties map[string]*sapcontrol.InstanceProperty `mapstructure:"properties,omitempty"`
+}
+
+func NewSAPSystemsList() (SAPSystemsList, error) {
+	var systems = SAPSystemsList{}
+
+	instances, err := findSystems()
+	if err != nil {
+		return systems, errors.Wrap(err, "Error walking the path")
+	}
+
+	for _, i := range instances {
+		s, err := NewSAPSystem(i)
+		if err != nil {
+			log.Printf("Error discovering a SAP system: %s", err)
+			continue
+		}
+		systems = append(systems, &s)
+	}
+
+	return systems, nil
+}
+
+// The Id is a unique identifier of a SAP installation.
+// It will be used to create totally independent SAP system data
+func (s *SAPSystem) getSapSystemId() (string, error) {
+	sid := s.Properties["SAPSYSTEMNAME"].Value
+	return internal.Md5sum(fmt.Sprintf("/usr/sap/%s/SYS/global/security/rsecssfs/key/SSFS_%s.KEY", sid, sid))
+}
+
+// Find the installed SAP instances in the /usr/sap folder
+func findSystems() ([]string, error) {
+	var instances = []string{}
+
+	_, err := os.Stat(sapInstallationPath)
+	if os.IsNotExist(err) {
+		log.Print("SAP installation not found")
+		return instances, nil
+	}
+
+	files, err := ioutil.ReadDir(sapInstallationPath)
+	if err != nil {
+		return nil, err
+	}
+
+	reSAPIdentifier := regexp.MustCompile(sapIdentifierPattern)
+
+	for _, f := range files {
+		if reSAPIdentifier.MatchString(f.Name()) {
+			log.Printf("New SAP system installation found: %s", f.Name())
+			i, err := findInstances(path.Join(sapInstallationPath, f.Name()))
+			if err != nil {
+				log.Print(err.Error())
+			}
+			instances = append(instances, i[:]...)
+		}
+	}
+
+	return instances, nil
+}
```

Comment:
This is done in purpose (at least in this particular case). When we have an error in the `findInstances` function, I didn't want to stop the whole finding process, so it just prints the error message but continues the process. 
It will just return an error if no SAP system is found. Otherwise, it will just return an empty list of instances within a SAP systems (I don't know if this happens, but maybe the user remove some instance, or change the user rights, or anything like that).

Reply:
hmm, I see; so, the only actual error condition in this function is an I/O error while reading the installation path?

Reply:
Hmm, yes, I would actually add the `continue` statement, now that you mention it: we should never rely on the return value of a function in presence of an error. It is also more communicative of the intent to continue looping even in presence of an error.